### ISSUE TITLE
Collections of notebooks

### ISSUE THREAD

**Issue by nthiery** [OpenDreamKit/OpenDreamKit#272]
# How to host collections of notebooks?

Many of us have had the need to collect and publish on the web collections of notebooks, e.g. for a class, or around a software package or a theme, like Mike's [collection of interacts](https://mikecroucher.github.io/jupyter-interactions/). ODK's website itself should host / link to such collections:
- demo notebooks for all our software
- demo notebooks illustrating our achievements
- demo notebooks by theme (e.g. around JOOMF, course notes, ...)
- ...

A Git repo + binder goes already a long way. But what about an notebook thumbnails to get an overview of the available notebooks at a glance? Search? Ranking? Simpler way to contributes notebooks? Metadata handling?

# Two new solutions

There are several projects around, like [nbgallery](https://nbgallery.github.io/) that try to address such needs. As usual, there is no one size fits all solution; going the ODK way and for our deliverable D2.7 (#47),  @defeo has kickstarted a modular toolbox for such needs called [planetaryum](https://github.com/OpenDreamKit/planetaryum). Currently the bits can be assembled to glorify a git repo with notebooks as a static web site. Features: rendering of the notebooks, links to binder, index page with automatically generated notebook thumbnails; soon: search capabilities; contributions by PR's and automatic update of the web site by continuous integration. See the [demo](https://opendreamkit.org/planetaryum-example-static/) and its [repo](https://github.com/OpenDreamKit/planetaryum-example-static).

Meanwhile, @sebasguts has written some Jekyll code for [Oscar's](http://oscar.computeralgebra.de/) and [ODK](http://opendreamkit.org)'s web site  that makes it easy to create a collection by providing a list of links to notebooks hosted on other repositories + metadata (see this  [PR](https://github.com/OpenDreamKit/OpenDreamKit.github.io/pull/83). Features: index page with (manually generated) thumbnails, binder links. Additional feature: the notebook hosting and binder configuration are outsourced to the external repos. See this [demo](https://opendreamkit.org/try/try/) and the underlying [metadata](https://github.com/OpenDreamKit/OpenDreamKit.github.io/blob/master/_data/try.yml). 

# What needs to be done

## A collection of demo notebook for the software components in ODK 

- [ ] Decide on the URL: http://opendreamkit.org/try/software (?)
- [ ] Use planetarium or links to remote notebooks?
- [ ] Each software (GAP, Pari, ...): provide a preferred demo notebook
- [ ] Link from ODK's main page and advertise

## A collection of demo notebooks for ODK

- [ ] Decide on the URL: http://opendreamkit.org/try/examples (?)
      Or should it be on the same page odk/try as above, just in a second section?
- [ ] Use planetarium or links to remote notebooks?
- [ ] Everyone: contribute nice notebooks
- [ ] Link from ODK's main page and advertise

## Create other collections of notebooks for ODK

- [ ] ???
- [ ] ???

## Create your own collections of notebooks in or out ODK

Try the above solutions, report back, contribute back, ...

## Advertise the solutions

**Comment by jdemeyer** [comment]
I have a test notebook for interacts containing interacts from https://wiki.sagemath.org/interact ported to Jupyter.

**Comment by defeo** [comment]
> I have a test notebook for interacts containing interacts from https://wiki.sagemath.org/interact ported to Jupyter.

Exactly what I needed! Thanks.

Do you know if they work in http://nbviewer.jupyter.org/ ?

**Comment by nthiery** [comment]
> About quantecon and Bookshelf.

Good to know! @defeo: I added a TODO about mentioning Bookshelf in the report.

**Comment by nthiery** [comment]
It's great to have Sage's interact collection as a notebook indeed!

Just random thoughts on how to best publish it:

- Should it take the form of a single notebook? or rather a collection
  of notebooks, one per interact? The latter form would enable
  browsing the interacts by thumbnails in the index page. But it
  presumably makes it slower to navigate from one interact to the
  other because of the page/notebook change.

- The interacts are sorted in sections on the wiki page; we could
  mimic this with subdirectories. Now, does/will planetaryum handle
  subdirectories, e.g. with folding sections in the index?

  or maybe the sections should be modeled by tags ...

**Comment by jdemeyer** [comment]
> Do you know if they work in http://nbviewer.jupyter.org/ ?

That renders static HTML pages, which is a bit pointless for interacts.

**Comment by jdemeyer** [comment]
> Should it take the form of a single notebook?

No. It's way to big for that. Maybe one notebook per section on that wiki page.

**Comment by defeo** [comment]
> Do you know if they work in http://nbviewer.jupyter.org/ ?
>
> That renders static HTML pages, which is a bit pointless for interacts.

Don't ipywidgets have some sort of static fallback? I see there's some
code for widgets in nbconvert exports:

https://github.com/jupyter/nbconvert/blob/master/nbconvert/templates/html/full.tpl#L17-L44

but I don't know what it is supposed to do, and I haven't seen an
example where this does anything.

**Comment by nthiery** [comment]
>    Don't ipywidgets have some sort of static fallback? I see there's some
>    code for widgets in nbconvert exports:
>    https://github.com/jupyter/nbconvert/blob/master/nbconvert/templates/ht
>    ml/full.tpl#L17-L44
>    but I don't know what it is supposed to do, and I haven't seen an
>    example where this does anything.

I indeed remember witnessing work to make widgets store their
javascript-side state in the notebook so that, when converting ditto
notebooks to e.g. static html+js, the widgets would still be visible
and functional (at least for anything not requiring kernel callbacks).
But I don't remember the precise state of affairs. @SylvainCorlay
would know.

**Comment by vidartf** [comment]
Embedded widgets can be interactive, but cannot do anything that needs the kernel (e.g. interact). It will simply show the state that the widget was in when the state was embedded in the notebook. Any widgets that are coupled with a `jslink` will still talk to each other though!

**Comment by trallard** [comment]
Here are some links to notebooks generated within ODK:
- [https://github.com/trallard/BAD_days](https://github.com/trallard/BAD_days)
- [https://github.com/trallard/ReproduciblePython](https://github.com/trallard/ReproduciblePython)
- [https://github.com/trallard/COM3240_Adaptive_Intelligence](https://github.com/trallard/COM3240_Adaptive_Intelligence)
- [https://github.com/trallard/WebData_Python](https://github.com/trallard/WebData_Python)
- [https://github.com/trallard/WebData_Python](https://github.com/trallard/WebData_Python)

**Comment by nthiery** [comment]
>    Here are some links to notebooks generated within ODK:

Excellent. Thank you!
